modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -264,50 +277,168 @@ struct IEUnit {
     }
 };
 
-struct IECallContext
+class IECallContext
 {
-    // Input parameters passed to an inference operation.
-    std::vector<cv::GArg> args;
-    cv::GShapes in_shapes;
+public:
+    IECallContext(const cv::GArgs                                   &  args,
+                  const std::vector<cv::gimpl::RcDesc>              &  outs,
+                  std::vector<cv::gimpl::GIslandExecutable::InObj>  && input_objs,
+                  std::vector<cv::gimpl::GIslandExecutable::OutObj> && otuput_objs);
 
-    //FIXME: avoid conversion of arguments from internal representation to OpenCV one on each call
-    //to OCV kernel. (This can be achieved by a two single time conversions in GCPUExecutable::run,
-    //once on enter for input and output arguments, and once before return for output arguments only
-    //FIXME: check if the above applies to this backend (taken from CPU)
-    std::unordered_map<std::size_t, cv::GRunArgP> results;
+    const cv::GArgs& inArgs() const;
 
     // Generic accessor API
     template<typename T>
-    const T& inArg(std::size_t input) { return args.at(input).get<T>(); }
+    const T& inArg(std::size_t input) const {
+        return m_args.at(input).get<T>();
+    }
 
-    const cv::MediaFrame& inFrame(std::size_t input) {
-        return inArg<cv::MediaFrame>(input);
+    template<typename T>
+     std::vector<T>& outVecR(std::size_t output) {
+        return outVecRef(output).wref<T>();
     }
 
     // Syntax sugar
-    const cv::Mat&   inMat(std::size_t input) {
-        return inArg<cv::Mat>(input);
-    }
-    cv::Mat&         outMatR(std::size_t output) {
-        return *cv::util::get<cv::Mat*>(results.at(output));
+          cv::GShape      inShape(int i)             const;
+    const cv::Mat&        inMat(std::size_t input)   const;
+    const cv::MediaFrame& inFrame(std::size_t input) const;
+
+    cv::Mat&     outMatR(std::size_t output);
+    cv::GRunArgP out(int output);
+
+private:
+    cv::detail::VectorRef& outVecRef(std::size_t output);
+
+    cv::GArg packArg(const cv::GArg &arg);
+
+    // To store input/output data from frames
+public:
+    std::vector<cv::gimpl::GIslandExecutable::InObj>  m_input_objs;
+    std::vector<cv::gimpl::GIslandExecutable::OutObj> m_output_objs;
+
+    // To simplify access to cv::Mat inside cv::RMat
+    cv::gimpl::Mag m_res;
+
+    // FIXME: avoid conversion of arguments from internal representation to OpenCV one on each call
+    //to OCV kernel. (This can be achieved by a two single time conversions in GCPUExecutable::run,
+    //once on enter for input and output arguments, and once before return for output arguments only
+    // FIXME: check if the above applies to this backend (taken from CPU)
+    std::unordered_map<std::size_t, cv::GRunArgP> m_results;
+
+    // Input parameters passed to an inference operation.
+    cv::GArgs m_args;
+    cv::GShapes m_in_shapes;
+
+    // Task additional params
+    cv::GArg state;
+    std::mutex m;
```

Comment:
Impossible to store sync primitives inside of `state`.  Is there more graceful approach ?